parser.go

```diff
@@ -0,0 +1,100 @@
+
+package cassandra
+
+import (
+	"fmt"
+	"strings"
+
+	sdk "github.com/conduitio/conduit-connector-sdk"
+)
+
+const (
+	insertQuery = "INSERT INTO %s (%s) VALUES (%s)"
+	updateQuery = "UPDATE %s SET %s WHERE %s"
+	deleteQuery = "DELETE FROM %s WHERE %s"
+
+	setStatementSeparator   = ","
+	whereStatementSeparator = "AND"
+)
+
+// Parser parses a record into a CQL query statement and its values
+type Parser struct{}
+
+// BuildInsertQuery takes a record, and returns the insert query statement and values representing that record.
+func (p *Parser) BuildInsertQuery(rec sdk.Record, table string) (string, []interface{}) {
+	keyCols, keyVals, cols, vals := p.getColumnsAndValues(rec.Key.(sdk.StructuredData), rec.Payload.After.(sdk.StructuredData))
+	cols = append(cols, keyCols...)
+	vals = append(vals, keyVals...)
+	query := fmt.Sprintf(insertQuery, table, strings.Join(cols, ", "), p.getPlaceholders(len(cols)))
+	return query, vals
+}
+
+// BuildUpdateQuery takes a record, and returns the update query statement and values representing that record.
+func (p *Parser) BuildUpdateQuery(rec sdk.Record, table string) (string, []interface{}) {
+	keyCols, keyVals, cols, vals := p.getColumnsAndValues(rec.Key.(sdk.StructuredData), rec.Payload.After.(sdk.StructuredData))
+	setStatement := p.pairValuesWithPlaceholder(cols, setStatementSeparator)
+	whereStatement := p.pairValuesWithPlaceholder(keyCols, whereStatementSeparator)
+	vals = append(vals, keyVals...)
+	query := fmt.Sprintf(updateQuery, table, setStatement, whereStatement)
+	return query, vals
+}
+
+// BuildDeleteQuery takes a record, and returns the delete query statement and values representing that record.
+func (p *Parser) BuildDeleteQuery(rec sdk.Record, table string) (string, []interface{}) {
+	keyCols, keyVals, _, _ := p.getColumnsAndValues(rec.Key.(sdk.StructuredData), rec.Payload.After.(sdk.StructuredData))
+	whereStatement := p.pairValuesWithPlaceholder(keyCols, whereStatementSeparator)
+	query := fmt.Sprintf(deleteQuery, table, whereStatement)
+	return query, keyVals
+}
+
+// getPlaceholders returns a string of question marks seperated by a comma with a given length.
+func (p *Parser) getPlaceholders(length int) string {
+	return strings.TrimSuffix(strings.Repeat("?, ", length), ", ")
+}
+
+func (p *Parser) pairValuesWithPlaceholder(cols []string, separator string) string {
+	var result string
+	for i := 0; i < len(cols); i++ {
+		result += cols[i] + " = ? " + separator + " "
+	}
+
+	// remove the trailing separator from the resulting string
+	result = strings.TrimSuffix(result, " "+separator+" ")
+
+	return result
+}
+
+// getColumnsAndValues returns the key columns and values, and the payload columns and values, each in a slice and in the order mentioned.
+func (p *Parser) getColumnsAndValues(key, payload sdk.StructuredData) ([]string, []interface{}, []string, []interface{}) {
+	var keyColumns []string
+	var keyValues []interface{}
+	var columns []string
+	var values []interface{}
```

Comment:
Nitpick: We have a pretty good idea of how much data we want to store in these slices, we can prepare them in advance with the correct capacity and potentially avoid unnecessary allocations (performance++ 🙂).
```suggestion
	keyColumns := make([]string, 0, len(key))
	keyValues := make([]interface{}, 0, len(key))
	columns := make([]string, 0, len(key))
	values := make([]interface{}, 0, len(key))
```

Reply:
columns and values should be `len(payload)` .. but got the point! thanks for the suggestion